### ISSUE TITLE
feat: user roles and permissions

### ISSUE THREAD

**Issue by felangel** [shorebirdtech/shorebird#1935]
As a user, I want to be able to add collaborators who have different permission levels. For example, I may want to have a single user who has permission to publish releases and patches but have other users have read-only access to view releases, patches, insights, and be able to preview an patch locally.

Currently, if user adds a collaborator, that collaborator automatically gets access to view and modify an app (including managing releases and patches).

**Comment by btrautmann** [comment]
To anchor this to our ideal setup, I'd describe our end goal as the following:

- _Only_ CI is allowed to publish patches (or more generally manipulate patches assuming deactivating them etc. can be done via CLI). This feels like a "patch manager" role or similar. This would likely look like a token that's stored as a secret and read by shorebird during a publish.
- Collaborators (i.e developers) have a "developer" role that allows them to view anything in the console as noted in OP.

**Comment by felangel** [comment]
Related issues
- https://github.com/shorebirdtech/shorebird/issues/1433
- https://github.com/shorebirdtech/shorebird/issues/1337
- https://github.com/shorebirdtech/shorebird/issues/739

**Comment by btrautmann** [comment]
@felangel if you feel this has too much overlap with existing issues, feel free to close it as such. I know this sorta spawned out of conversation with myself and @ClaireDavis--neither of us will be offended if we need to subscribe to another issue 😄

**Comment by eseidel** [comment]
Yeah, more issues isn't a problem. We know we need to do a bunch of work in this area, we've just been a) really busy with iOS and 1.0 and b) waiting until we had more reports from customers (like this one!) so we know we're building the right thing.  Thank you!

**Comment by eseidel** [comment]
This ends up part of https://github.com/shorebirdtech/shorebird/issues/1433 which I think is probably our next big area of work after iOS performance (maybe we'll do asset support before that?)

**Comment by eseidel** [comment]
Much of this is launched now as part of https://github.com/shorebirdtech/shorebird/issues/739.

Will keep this bug open for adding further permission levels.

We had an explicit request today from a large customer to separate "create a patch" from "send a patch to users" roles.

**Comment by eseidel** [comment]
Specifically the "use Shorebird without being able to deploy a patch to users" is the main missing role in enterprise conversations.

**Comment by eseidel** [comment]
@bryanoltman is working on upgrading our RBAC infrastructure now which will make a "viewer" role (and other more specific roles) possible.  After he's done we will close this and open any remaining more specific issues.  Thanks.

**Comment by eseidel** [comment]
I've committed to a customer we will deliver the current RBAC work by end of Dec 2025.

**Comment by eseidel** [comment]
We've launched this: https://shorebird.dev/blog/advanced-permissions/
